packages/chrome-offscreen-worker/src/control.ts

```diff
@@ -0,0 +1,160 @@
+    }
+    this.offscreenPath = offscreen.href;
+  }
+
+  private async createOffscreen(): Promise<OffscreenRootPort> {
+    const {
+      promise: offscreenConnection,
+      resolve: resolveConnection,
+      reject: rejectConnection,
+    } = Promise.withResolvers<chrome.runtime.Port>();
+
+    const hasOffscreen = async () => {
+      console.log('hasOffscreen');
+      const contexts = await chrome.runtime.getContexts({
+        contextTypes: [chrome.runtime.ContextType.OFFSCREEN_DOCUMENT],
+      });
+      console.log('hasOffscreen contexts', contexts);
+      const found = contexts.find(ctx => {
+        console.log('contexts.find', ctx.documentUrl, this.offscreenPath);
+        return ctx.documentUrl?.startsWith(this.offscreenPath);
+      });
+      console.log('hasOffscreen found', found);
+      return found;
+    };
+
+    let offscreenHash: string | undefined;
+    try {
+      const offscreenContext = await hasOffscreen();
+      if (!offscreenContext?.documentUrl) {
+        console.log('createOffscreen !offscreenContext');
+        const offscreenInitUrl = new URL(`#${crypto.randomUUID()}`, this.offscreenPath);
+        console.log('initting', offscreenInitUrl.href);
+        await chrome.offscreen.createDocument({
+          url: offscreenInitUrl.href,
+          reasons: [chrome.offscreen.Reason.WORKERS],
+          justification:
+            'Manages workers for parallelizing heavy computation outside of the service worker thread',
+        });
+        offscreenHash = offscreenInitUrl.hash;
+        console.log('createOffscreen created offscreenHash', offscreenInitUrl, offscreenHash);
+      } else {
+        offscreenHash = new URL(offscreenContext.documentUrl).hash;
+        console.log('createOffscreen located offscreenHash', offscreenHash);
+      }
+    } catch (creationFailure) {
+      console.warn(
+        "Failed to create offscreen document, you'll probably time out now",
+        creationFailure,
+        globalThis.location,
+      );
+      const offscreenContext = offscreenHash ? undefined : await hasOffscreen();
+      if (offscreenContext?.documentUrl) {
+        console.warn('Found offscreen context after failure', offscreenContext.documentUrl);
+        if (offscreenHash) {
+          console.warn('Discarding due to present', offscreenHash);
+        }
+        offscreenHash ??= new URL(offscreenContext.documentUrl).hash;
+      }
+    } finally {
+      const attempt = offscreenHash ?? 'no hash';
+      attempts.push(attempt);
+      console.warn('attempt', attempts.length, attempt);
+    }
+
+    if (!offscreenHash) {
+      rejectConnection('No offscreen context');
+    } else {
+      const sessionPort = chrome.runtime.connect({ name: offscreenHash });
+      sessionPort.onDisconnect.addListener(() => {
+        this.workers.forEach(p => p.disconnect());
+        this.workers.clear();
+      });
+      resolveConnection(sessionPort);
+    }
+
+    return offscreenConnection;
+  }
```

Comment:
this method will be refactored, as it presently tries to begin using the document too quickly.